<a href="https://colab.research.google.com/github/vtiha/Tytorenko_Assignments/blob/main/9_2_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment 9.2

> Replace all TODOs with your code. Do not change any other code.

In [2]:
# Do not edit this cell

from typing import List

## Descriptive statistics

In this assignment, we will write the functions to calculate the basic statistics from scratch, not using numpy.

### Task 1

Let's start simple: write a function `mean` that calculates the average of the list.

$$\mu = \frac{{\sum_{i=1}^n x_i}}{{n}}$$

In [5]:
def mean(li: List[float]) -> float:
    return sum(li) / len(li)


assert mean([1., 2., 3.]) == 2.
assert mean([1., 1., 2., 0.]) == 1.

### Task 2

Now let's calculate variance (dispersion). You may use the `mean` function implemented before.

$$V = \frac{{\sum_{i=1}^n (x_i - \mu)^2}}{{n}}$$

In [6]:
def variance(li: List[float]) -> float:
    avg = mean(li)
    return sum((x - avg) ** 2 for x in li) / len(li)

assert variance([1., 1., 1.]) == 0.
assert variance([1., 2., 3., 4.]) == 1.25

### Task 3

The standard deviation is easy once you get the variance:

$$\sigma = \sqrt{V}$$

In [8]:
def std(li: List[float]) -> float:
    return (variance(li)) ** 0.5

assert std([1., 1., 1.]) == 0.
assert std([1., 2., 3., 4.]) == 1.25**0.5

### Task 4

**Median**

The median is the middle value in a sorted dataset. If the dataset has an odd number of values, the median is the value at the center. If the dataset has an even number of values, the median is the average of the two middle values.

In [10]:
def median(li: List[float]) -> float:
    li_sorted = sorted(li)
    n = len(li_sorted)

    if n % 2 == 1:
        return li_sorted[n // 2]
    else:
        return (li_sorted[n // 2 - 1] + li_sorted[n // 2]) / 2


assert median([1., 1., 1.]) == 1.
assert median([1., 4., 3., 2.]) == 2.5

## Measure performance

Sometimes, apart from theoretical, algorithmic complexity, it's a good idea to compare the runtime of two algorithms empirically, i.e., run the code many times and time it.

In Python's standard library, we have [timeit](https://docs.python.org/3/library/timeit.html) module that does exactly that.

Let's compare the runtime of your implementations and numpy. Use the provided setup code:

In [17]:
# generate data for tests
setup = '''
import random
import numpy as np

arr = np.random.rand(10_000) * 100
li = [random.random() * 100 for _ in range(10_000)]
'''

# pass your function to timeit module
funcs = {
    'mean': mean,
    'variance': variance,
    'std': std,
    'median': median,
}

### Task 5

Complete Python statements to compare your functions to numpy. Use `li` for your function and `arr` for numpy functions.

In [18]:
stmt_mean_custom = 'mean(li)'
stmt_mean_np = 'np.mean(arr)'

stmt_var_custom = 'variance(li)'
stmt_var_np = 'np.var(arr)'

stmt_std_custom = 'std(li)'
stmt_std_np = 'np.std(arr)'

stmt_median_custom = 'median(li)'
stmt_median_np = 'np.median(arr)'

### Task 6

Measure average exec time of your statements with `timeit` module. As your submission, fill out the table with results (rounded to 2 decimal places)

In [24]:
import timeit


mean_custom_time = timeit.timeit(stmt='mean(li)', setup=setup, globals=globals(), number=10_000)
mean_np_time = timeit.timeit(stmt='np.mean(arr)', setup=setup, globals=globals(), number=10_000)

var_custom_time = timeit.timeit(stmt='variance(li)', setup=setup, globals=globals(), number=10_000)
var_np_time = timeit.timeit(stmt='np.var(arr)', setup=setup, globals=globals(), number=10_000)

std_custom_time = timeit.timeit(stmt='std(li)', setup=setup, globals=globals(), number=10_000)
std_np_time = timeit.timeit(stmt='np.std(arr)', setup=setup, globals=globals(), number=10_000)

median_custom_time = timeit.timeit(stmt='median(li)', setup=setup, globals=globals(), number=10_000)
median_np_time = timeit.timeit(stmt='np.median(arr)', setup=setup, globals=globals(), number=10_000)

print(f"mean (custom): {mean_custom_time:.2f} secs")
print(f"mean (numpy): {mean_np_time:.2f} secs")

print(f"variance (custom): {var_custom_time:.2f} secs")
print(f"variance (numpy): {var_np_time:.2f} secs")

print(f"std (custom): {std_custom_time:.2f} secs")
print(f"std (numpy): {std_np_time:.2f} secs")

print(f"median (custom): {median_custom_time:.2f} secs")
print(f"median (numpy): {median_np_time:.2f} secs")

mean (custom): 0.48 secs
mean (numpy): 0.09 secs
variance (custom): 17.62 secs
variance (numpy): 0.35 secs
std (custom): 16.28 secs
std (numpy): 0.35 secs
median (custom): 16.50 secs
median (numpy): 1.28 secs


Time per 10000 executions, secs

| Func       | Custom | Numpy |
| ---------- | ------ | ----- |
| mean       |  0.48  | 0.09  |
| var        | 17.62  | 0.35  |
| std        | 16.28  | 0.35  |
| median     | 16.50  | 1.28  |